src/js/relayclient/ServerHelper.js

```diff
@@ -186,51 +201,92 @@ class ServerHelper {
       await this.fetchRelaysAdded()
     }
-    return this.createActiveRelayPinger(this.filteredRelays, this.httpSend, gasPrice, this.verbose)
+    return this.createActiveRelayPinger(this.filteredRelays, this.httpSend, gasPrice, this.verbose, this.preferredRelays)
   }
 
-  createActiveRelayPinger (filteredRelays, httpSend, gasPrice, verbose) {
-    return new ActiveRelayPinger(filteredRelays, httpSend, gasPrice, verbose)
+  createActiveRelayPinger (filteredRelays, httpSend, gasPrice, verbose, preferredRelays) {
+    let pinger = new ActiveRelayPinger(filteredRelays, httpSend, gasPrice, verbose)
+    if (preferredRelays) {
+      const prefs = this.preferredRelays.map(relayUrl => ({
+        relayUrl,
+        transactionFee: 0,
+        stake: PREFERRED_STAKE,
+        unstakeDelay: PREFERRED_DELAY
+      })
+      )
+      // if there are preferred relays, create a pinger that uses them FIRST, and only
+      // falls-back to the "normal" pinger, based on the filtered list.
+      pinger = new ActiveRelayPinger(prefs, httpSend, gasPrice, verbose, pinger)
+    }
+    return pinger
   }
 
   /**
    * Iterates through all RelayAdded and RelayRemoved logs emitted by given hub
    * initializes an array {@link filteredRelays} of relays curently registered on given RelayHub contract
    */
   async fetchRelaysAdded () {
-    const activeRelays = {}
+    this.verbose = true
     const fromBlock = this.fromBlock || 2
-    const addedAndRemovedEvents = await this.relayHubInstance.getPastEvents('allEvents', {
-      fromBlock: fromBlock
-      // topics: [["RelayAdded", "RelayRemoved"]]
+    const eventTopics = event2topic(this.relayHubInstance,
+      ['RelayAdded', 'RelayRemoved', 'TransactionRelayed', 'CanRelayFailed'])
+
+    const relayEvents = await this.relayHubInstance.getPastEvents('allEvents', {
+      fromBlock: fromBlock,
+      topics: [eventTopics]
     })
 
     if (this.verbose) {
-      console.log('fetchRelaysAdded: found ' + addedAndRemovedEvents.length + ' events')
+      console.log('fetchRelaysAdded: found ', relayEvents.length + ' events')
     }
-    // TODO: better filter RelayAdded, RelayRemoved events: otherwise, we'll be scanning all TransactionRelayed too...
-    // since RelayAdded can't be called after RelayRemoved, its OK to scan first for add, and the remove all removed relays.
-    for (var index in addedAndRemovedEvents) {
-      const event = addedAndRemovedEvents[index]
-      if (event.event === 'RelayAdded') {
-        const args = event.returnValues
-        const relay = {
-          address: args.relay,
-          relayUrl: args.url,
-          transactionFee: args.transactionFee,
-          stake: args.stake,
-          unstakeDelay: args.unstakeDelay
-        }
-        relay.score = this.calculateRelayScore(relay)
-        activeRelays[args.relay] = relay
-      } else if (event.event === 'RelayRemoved') {
-        delete activeRelays[event.returnValues.relay]
+    const foundRelays = {}
```

Comment:
There is a 'Set' built-in class, makes it more clear what this variable is

Reply:
```
> const s = new Set()
Set {}
> s.add("aaa")
Set { 'aaa' }
> s.add("aaa")
Set { 'aaa' }
> s.add("aaaa")
Set { 'aaa', 'aaaa' }
> s.delete('aaa')
true
> s
Set { 'aaaa' }

Reply:
good idea.